### PR TITLE
Added mhlo.dot to mhlo.mul conversion

### PR DESCRIPTION
**Pull request by rsuderman**
It is possible that the mhlo.dot is representable as an mhlo.mul.
Convert to that mult whenever possible.

### CODE REVIEW

File: compiler/src/iree/compiler/InputConversion/MHLO/MHLOToMHLOPreprocessing.cpp [hunk 1]
```diff
@@ -778,6 +778,69 @@ struct FuseWidenOperands : public OpRewritePattern<Op> {
...
 };
 
+struct DotToMul : public OpRewritePattern<mhlo::DotOp> {
+  using OpRewritePattern<mhlo::DotOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(mhlo::DotOp op,
+                                PatternRewriter &rewriter) const override {
+    auto lhs = op.lhs();
+    auto rhs = op.rhs();
+    auto lhsTy = lhs.getType().cast<RankedTensorType>();
```
Comment (jpienaar) [inline compiler/src/iree/compiler/InputConversion/MHLO/MHLOToMHLOPreprocessing.cpp:788]: This cast may assert/not required for op AFAIK. Dyn_cast and notifyMatchFailure?

File: compiler/src/iree/compiler/InputConversion/MHLO/MHLOToMHLOPreprocessing.cpp [hunk 2]
```diff
@@ -778,6 +778,69 @@ struct FuseWidenOperands : public OpRewritePattern<Op> {
...
+    auto rhsTy = rhs.getType().cast<RankedTensorType>();
+    auto resultTy = op.getType().cast<RankedTensorType>();
+
+    if (lhsTy.getDimSize(1) != 1) return failure();
+
+    Value batchSize = rewriter.create<mhlo::GetDimensionSizeOp>(
```
Comment (jpienaar) [inline compiler/src/iree/compiler/InputConversion/MHLO/MHLOToMHLOPreprocessing.cpp:794]: Could you add a high level description here?

File: compiler/src/iree/compiler/InputConversion/MHLO/MHLOToMHLOPreprocessing.cpp [hunk 3]
```diff
@@ -778,6 +778,69 @@ struct FuseWidenOperands : public OpRewritePattern<Op> {
...
+
+    Value featureSize = rewriter.create<mhlo::GetDimensionSizeOp>(
+        op.getLoc(), RankedTensorType::get({1}, rewriter.getI32Type()), rhs,
+        rewriter.getI64IntegerAttr(1));
+
+    Value outSize = rewriter.create<mhlo::ConcatenateOp>(
+        op.getLoc(), RankedTensorType::get({2}, rewriter.getI32Type()),
+        ValueRange{batchSize, featureSize}, rewriter.getI64IntegerAttr(0));
+
+    /*
```
Comment (jpienaar) [inline compiler/src/iree/compiler/InputConversion/MHLO/MHLOToMHLOPreprocessing.cpp:806]: I'd rather delete commented out code.

**Review by jpienaar — COMMENTED** [review]
Nice

File: compiler/src/iree/compiler/InputConversion/MHLO/MHLOToMHLOPreprocessing.cpp (hunk 1 above)
Comment (rsuderman) [reply compiler/src/iree/compiler/InputConversion/MHLO/MHLOToMHLOPreprocessing.cpp]: Added.

File: compiler/src/iree/compiler/InputConversion/MHLO/MHLOToMHLOPreprocessing.cpp (hunk 2 above)
Comment (rsuderman) [reply compiler/src/iree/compiler/InputConversion/MHLO/MHLOToMHLOPreprocessing.cpp:805]: Done.

File: compiler/src/iree/compiler/InputConversion/MHLO/MHLOToMHLOPreprocessing.cpp (hunk 3 above)
Comment (rsuderman) [reply compiler/src/iree/compiler/InputConversion/MHLO/MHLOToMHLOPreprocessing.cpp]: Deleted

File: compiler/src/iree/compiler/InputConversion/MHLO/MHLOToMHLOPreprocessing.cpp [hunk 4]
```diff
@@ -778,6 +778,68 @@ struct FuseWidenOperands : public OpRewritePattern<Op> {
...
+      return rewriter.notifyMatchFailure(op, "lhs and rhs must be ranked");
+    }
+
+    if (lhsTy.getRank() != 2 || rhsTy.getRank() != 2) {
+      return rewriter.notifyMatchFailure(op, "lhs and rhs must be rank-2");
+    }
+
+    if (lhsTy.getDimSize(1) != 1) return failure();
+
+    // Dynamically compoute the shape of the result of the DotOp by querying
```
Comment (jpienaar) [inline compiler/src/iree/compiler/InputConversion/MHLO/MHLOToMHLOPreprocessing.cpp:802]: compute

File: compiler/src/iree/compiler/InputConversion/MHLO/MHLOToMHLOPreprocessing.cpp [hunk 5]
```diff
@@ -778,6 +778,68 @@ struct FuseWidenOperands : public OpRewritePattern<Op> {
...
+    // the 0-th dimesnions of hte left, the 1st dimension of the right, and
```
Comment (jpienaar) [inline compiler/src/iree/compiler/InputConversion/MHLO/MHLOToMHLOPreprocessing.cpp:803]: dimensions, the

**Review by jpienaar — APPROVED** [review]
Thanks!

File: compiler/src/iree/compiler/InputConversion/MHLO/MHLOToMHLOPreprocessing.cpp (hunk 4 above)
Comment (rsuderman) [reply compiler/src/iree/compiler/InputConversion/MHLO/MHLOToMHLOPreprocessing.cpp]: Done.

File: compiler/src/iree/compiler/InputConversion/MHLO/MHLOToMHLOPreprocessing.cpp (hunk 5 above)
Comment (rsuderman) [reply compiler/src/iree/compiler/InputConversion/MHLO/MHLOToMHLOPreprocessing.cpp]: Done.

**Pull request merged**